robot/lib/Src/drivers/ICM20948.cpp

```diff
@@ -0,0 +1,193 @@
+    bool failed_init = false;
+
+    if (whoami != WHOAMI_VAL) {
+        printf("Failed to connect to IMU.");
+        failed_init = true;
+    }
+
+    // Reset
+    write_register(0, Registers::PWR_MGMT_1, 0xC1);
+
+    HAL_Delay(100);
+
+    // Check whoami again
+    whoami = read_register(0, Registers::WHO_AM_I);
+
+    if (whoami != WHOAMI_VAL && failed_init) {
+        printf("Failed to connect to IMU.");
+        return false;
+    }
+    
+    // Auto select best clock source
+    write_register(0, Registers::PWR_MGMT_1, 0x01);
+    
+    // Operate I2C in duty cycled mode and disable GYRO duty cycle mode
+    write_register(0, Registers::LP_CONFIG, 0x40);
+    
+    // Disable I2C
+    write_register(0, Registers::USER_CTRL, 0x10);
+
+    // Turn FIFO off
+    write_register(0, Registers::FIFO_MODE, 0x00);
+
+    // Enable gyro
+    write_register(0, Registers::PWR_MGMT_2, 0x00);
+    
+    // Configure gyro
+    write_register(2, Registers::GYRO_CONFIG_1, 0x04);
+    
+    // Check whoami again
+    whoami = read_register(0, Registers::WHO_AM_I);
+
+    if (whoami != WHOAMI_VAL) {
+        printf("Failed to connect to IMU.");
+        return false;
+    } else {
+        printf("Initialization Complete!");
+        return true;
+    }
+}
+
+float ICM20948::gyro_z() {
+
+    uint16_t hi = read_register(0, Registers::GYRO_Z_OUT_H),
+             lo = read_register(0, Registers::GYRO_Z_OUT_L);
+
+
+    // Reinterpret the bits as a signed 16-bit integer
+    int16_t total = hi << 8 | lo;
+
+    // +-1000dps max, for 16-bit signed integers, comes out to 32.8lsb/deg.
+    const float lsbToDps = 1000.0f / 32768.f;
+    const float degToRad = M_PI / 180.0f;
+    return total * (degToRad * lsbToDps);
+}
+
+/*
+uint8_t ICM20948::read_register(uint8_t address) {
+    bool was_cs_active = currently_active;
+
+    chip_select(true);
+
+    uint8_t data_tx[2] = {address | READ_BIT, 0x0};
+    uint8_t data_rx[2] = {0x0, 0x0};
+    imuSPI->transmitReceive(data_tx, data_rx, 2);
+
+    printf("%x%x -> %x%x\r\n", data_tx[0], data_tx[1], data_rx[0], data_rx[1]);
+
+    // Restore previous CS state
+    chip_select(was_cs_active);
+
+    return data_rx[1];
+}
+
+void ICM20948::write_register(uint8_t address, uint8_t value) {
+    bool was_cs_active = currently_active;
+
+    chip_select(true);
+
+    uint8_t data_tx[2] = {address, value};
+    uint8_t data_rx[2] = {0x0, 0x0};
+    imuSPI->transmitReceive(data_tx, data_rx, 2);
+
+    // Restore previous CS state
+    chip_select(was_cs_active);
+}
+
+void ICM20948::chip_select(bool active) {
+    if (currently_active != active) {
+        currently_active = active;
+        chip_select_pin.write(active);
+    }
+}
+*/
+
+void ICM20948::chip_select(bool cs_state) {
+    if (cs_state) {
+        nCs.write(0);
+    } else {
+        nCs.write(1);
+    }
+}
+
+void ICM20948::write_register(uint8_t bank, uint8_t address, uint8_t value) {
+    chip_select(true);
```

Comment:
Only set bank select if it's changed from the last one set.

Reply:
Again, this.

Just record the latest bank selected (remember to reset it when resetting the device). This is going to cause a lot of extra SPI writes as-is (not the end of the world, just a little bit overkill).